blockchain/chain.go

```diff
@@ -228,6 +228,39 @@ func (b *BlockChain) HaveBlock(hash *chainhash.Hash) (bool, error) {
 	return exists || b.IsKnownOrphan(hash), nil
 }
 
+// HaveBlockWithData returns whether or not the chain instance has the block
+// represented by the passed hash This includes checking the various places
+// a block can be like part of the main chain, on a side chain, or in the orphan
+// pool. In addition to this the function also checks for presence of block
+// data along with presence of node.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) HaveBlockWithData(hash *chainhash.Hash) (bool, error) {
+	// Checking if node exists in blockIndex or database.
+	exists, err := b.blockExists(hash)
+	if err != nil {
+		return false, err
+	}
+	if exists || b.IsKnownOrphan(hash) {
+		// Retrieving node from the blockIndex so that we can check the node
+		// status.
+		node := b.IndexLookupNode(hash)
```

Comment:
I am checking again for the node in `IndexLookupNode` because of the case where the block is stored in the database and is not present in the Index.
In that case the returned node would be nil right?

Reply:
I got your point, that checking for block data for an orphan block has no point, since we can't have an orphan header.
So I can just return true if the block is an orphan.